File: simple/cppvulnweb/cmd_exec/sub_class_test.cpp
#include "common/vuln_child.h"
#include <iostream>
#include <string>

// 命令执行漏洞 - 纯子类继承调用测试文件
// 测试场景：通过子类 VulnChild 继承父类 VulnParent 的漏洞方法

void cmd_exec_sub_class_test(const std::string& cmd) {
    std::cout << "=== 命令执行漏洞 - 子类继承调用测试 ===" << std::endl;

    // 创建子类实例
    VulnChild child;

    // 场景1：外部可控数据源（通过请求参数传入）
    // 调用链：VulnChild -> VulnParent::executeCommand -> popen（漏洞点）
    std::string result1 = child.executeCommand(cmd);
    std::cout << "[外部参数] 命令执行结果: " << result1 << std::endl;

    // 场景2：硬编码无外部数据源
    std::string hardcoded_cmd = "echo hardcoded_cmd_exec_test";
    // 调用链：VulnChild -> VulnParent::executeCommand -> system（漏洞点）
    std::string result2 = child.executeCommand(hardcoded_cmd);
    std::cout << "[硬编码] 命令执行结果: " << result2 << std::endl;
}

File: simple/javavulnweb/src/main/java/com/vulnweb/cmd_exec/SubClassTest.java
package com.vulnweb.cmd_exec;

import com.vulnweb.common.VulnChild;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

/**
 * 命令执行漏洞 - 纯子类继承调用测试文件
 * 通过 VulnChild 子类调用父类漏洞方法
 */
@RestController
public class SubClassTest {
    // 创建子类实例
    private VulnChild child = new VulnChild();

    /**
     * 测试命令执行：硬编码命令
     * 调用子类方法，子类调用父类漏洞方法
     */
    @GetMapping("/cmd/sub/hardcoded")
    public String testHardcoded() {
        String result = child.executeCmd("echo subClassHardcoded");
        return "子类继承硬编码命令执行结果: " + result;
    }

    /**
     * 测试命令执行：外部参数
     * 接收前端参数，通过子类调用父类漏洞方法
     */
    @GetMapping("/cmd/sub/param")
    public String testParam(@RequestParam String cmd) {
        // 调用链：请求参数 -> SubClassTest -> VulnChild -> VulnParent.cmdExec
        String result = child.executeCmd(cmd);
        return "子类继承带参命令执行结果: " + result;
    }
}